### ISSUE TITLE
search with boolean OR not commutative when date range is involved

### ISSUE THREAD

**Issue by tsgit** [inveniosoftware/invenio#2364]
find (A ELLISON and AFF RIVERSIDE and TC P) or (AFF IMPERIAL and DATE <1989 and A ELLISON)
is not equivalent to 
find (AFF IMPERIAL and DATE <1989 and A ELLISON) or (A ELLISON and AFF RIVERSIDE and TC P)
the former is inclusive of 1989 (although it shouldn't be), the other isn't.

x1=perform_request_search(p="find (A ELLISON and AFF RIVERSIDE and TC P) or (AFF IMPERIAL and DATE <1989 and A ELLISON)")
x2=perform_request_search(p="find (AFF IMPERIAL and DATE <1989 and A ELLISON) or (A ELLISON and AFF RIVERSIDE and TC P)")
set(x1)-set(x2)
set([865867, 264955, 865868, 277701])
set(x2)-set(x1)
set([])

looking at the distribution of the OR over the parenthesized AND expressions, they are equivalent except for the "- year:1989" terms labelled 3) 7) 11) below

Search stage 1: search_pattern_parenthesised() searched '(author:ELLISON and
affiliation:RIVERSIDE and collection:P) or (affiliation:IMPERIAL and
(year:0->1989 AND NOT year:1989) and author:ELLISON)'.

Search stage 1: search_pattern_parenthesised() returned 
1)   ['+', 'affiliation:imperial | author:ellison', 
2)   '+', 'year:0->1989 | author:ellison', 
3)   '+', '- year:1989 | author:ellison', 
4)   '+', 'author:ellison | author:ellison', 
5)   '+', 'affiliation:imperial | affiliation:riverside', 
6)   '+', 'year:0->1989 | affiliation:riverside', 
7)   '+', '- year:1989 | affiliation:riverside', 
8)   '+', 'author:ellison | affiliation:riverside', 
9)   '+', 'affiliation:imperial | collection:p', 
10)  '+', 'year:0->1989 | collection:p',
11)  '+', '- year:1989  | collection:p', 
12)  '+', 'author:ellison | collection:p'].
     

Search stage 1: search_pattern_parenthesised() searched
'(affiliation:IMPERIAL and (year:0->1989 AND NOT year:1989) and
author:ELLISON) or (author:ELLISON and affiliation:RIVERSIDE and
collection:P)'.

Search stage 1: search_pattern_parenthesised() returned 
1)   ['+', 'author:ellison | affiliation:imperial', 
5)   '+', 'affiliation:riverside | affiliation:imperial',
9)   '+', 'collection:p | affiliation:imperial', 
2)   '+', 'author:ellison | year:0->1989', 
6)   '+', 'affiliation:riverside | year:0->1989', 
10)  '+', 'collection:p | year:0->1989', 
3)   '+', 'author:ellison - year:1989', 
7)   '+', 'affiliation:riverside - year:1989', 
11)  '+', 'collection:p - year:1989', 
4)   '+', 'author:ellison | author:ellison', 
8)   '+', 'affiliation:riverside | author:ellison', 
12)  '+', 'collection:p | author:ellison'].

and looking further into how e.g. 
11)  '+', 'collection:p - year:1989', 
 is interpreted as opposed to 
11)  '+', '- year:1989  | collection:p', 

the former leads to 
Search stage 1: basic search units are: [['+', 'p', 'collection', 'w'], ['-', '1989', 'year', 'w']]
while the latter gives
Search stage 1: basic search units are: [['+', '1989', 'year', 'w'], ['|', 'p', 'collection', 'w']]

what happened to the "-" in the latter?

This just based on observation of verbose output, didn't have time to inspect the code
Cheers
T.

**Comment by tiborsimko** [comment]
@kaplun Can you please check whether this is still the case?

**Comment by kaplun** [comment]
It's still the case today on production. I am checking if it's also a bug in `invenio_query_parser`

**Comment by Panos512** [comment]
I can confirm that the output of invenio-query-parser is correct: 
```
x.parse_query("find (a ELLISON and aff RIVERSIDE and tc P) or (aff IMPERIAL and d <1989 and a ELLISON)").accept(walker())

OrOp(
  AndOp(
    AndOp(
      KeywordOp(Keyword('author'), Value('ELLISON')), 
      KeywordOp(Keyword('affiliation'), Value('RIVERSIDE'))), 
    KeywordOp(Keyword('collection'), Value('P'))), 
  AndOp(
    AndOp(
      KeywordOp(Keyword('affiliation'), Value('IMPERIAL')), 
      KeywordOp(Keyword('year'), LowerOp(Value('1989')))), 
    KeywordOp(Keyword('author'), Value('ELLISON'))))


x.parse_query("find (aff IMPERIAL and d <1989 and a ELLISON) or (a ELLISON and aff RIVERSIDE and tc P)").accept(walker())

OrOp(
  AndOp(
    AndOp(
      KeywordOp(Keyword('affiliation'), Value('IMPERIAL')), 
      KeywordOp(Keyword('year'), LowerOp(Value('1989')))), 
    KeywordOp(Keyword('author'), Value('ELLISON'))), 
  AndOp(
    AndOp(
      KeywordOp(Keyword('author'), Value('ELLISON')), 
      KeywordOp(Keyword('affiliation'), Value('RIVERSIDE'))), 
    KeywordOp(Keyword('collection'), Value('P'))))
```
I will add these two as test cases for future testing.

**Comment by jirikuncar** [comment]
Addressed in inveniosoftware/invenio-query-parser#33

**Comment by kaplun** [comment]
Fixed in `invenio-query-parser`.
